Add JoinMarket Clientserver

nixbitcoin opened the pull request:
This PR integrates JoinMarket into nix-bitcoin. All user scripts are created with `pkgs.writeScriptBin` and are executed in the `cfg.dataDir` with the right user and options. `tumbler.py` can be started in the background using GNU Screen. Yieldgenerator can be run as a systemd service.

How could I firewall the entire joinmarket user to only use Tor?

Thank you @erikarvstedt for your essential help throughout packaging this complex python program.

Closes https://github.com/fort-nix/nix-bitcoin/issues/32

nixbitcoin commented:
Rebased

nixbitcoin commented:
Rebased and ready for review

nixbitcoin commented:
Updated to JoinMarket 0.7.0 & rebased. Integrated `netns-isolation` and `cliExec`.

I'm relying on manual wallet generation at the moment as described in `/docs/usage.md`. This has advantages if the user wants to customize his wallet experience, with/without 2FA or Fidelity Bonds for example. However, it also creates friction because the user must manually enter his passwords in the local secrets to use the `joinmarket-yieldgenerator` bot.

We could implement a low-level python script to automatically create a wallet and save the seed, as in our LND module. An example of such a python script can be found in https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/scripts/joinmarket-qt.py#L2010-L2018.

erikarvstedt commented:
[Fixups](https://github.com/erikarvstedt/nix-bitcoin/commits/joinmarket).

Regarding `python-bitcointx: fix provision of openssl dependency`: This is a good example on how nix differs from traditional packaging by making all runtime deps explicit. It's really helpful to understand why the previous approach didn't work.

I'm not very happy about `pkg: make bin script names more cli friendly`, but I think that's the best way to go for now. I've suggested an improved CLI [at the jm repo](https://github.com/JoinMarket-Org/joinmarket-clientserver/issues/681).

A script to automatically create a wallet would be a great idea, if only for automated testing.

Why does joinmarketd need privileged bitcoin RPC access?

erikarvstedt commented:
How would a user configure the yieldgenerator? Edit `joinmarket.cfg`? We should then add a `extraOptions` option.

erikarvstedt commented:
So joinmarket uses bitcoind wallet functionality? We should then set `bitcoind.disablewallet = false` and add a comment why have to use the privileged RPC user.

nixbitcoin commented:
`joinmarket: automatically generate wallet` now no longer WIP.

I've also added 0535da4. I was getting
```
Sep 12 18:18:50 nix-bitcoin bash[15886]: Amp server or network failure unhandled by client application.  Dropping connection!  To avoid, add errbacks to ALL remote commands!
Sep 12 18:18:50 nix-bitcoin bash[15886]: Traceback (most recent call last):
Sep 12 18:18:50 nix-bitcoin bash[15886]: --- <exception caught here> ---
Sep 12 18:18:50 nix-bitcoin bash[15886]:   File "/nix/store/6li72ddcdwcdg1m4zdghv2xjw6a72sf7-python3-3.7.6-env/lib/python3.7/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
Sep 12 18:18:50 nix-bitcoin bash[15886]:     current.result = callback(current.result, *args, **kw)
Sep 12 18:18:50 nix-bitcoin bash[15886]:   File "/nix/store/6li72ddcdwcdg1m4zdghv2xjw6a72sf7-python3-3.7.6-env/lib/python3.7/site-packages/jmclient/client_protocol.py", line 51, in defaultErrback
Sep 12 18:18:50 nix-bitcoin bash[15886]:     ConnectionLost) #pragma: no cover
Sep 12 18:18:50 nix-bitcoin bash[15886]:   File "/nix/store/6li72ddcdwcdg1m4zdghv2xjw6a72sf7-python3-3.7.6-env/lib/python3.7/site-packages/twisted/python/failure.py", line 460, in trap
Sep 12 18:18:50 nix-bitcoin bash[15886]:     self.raiseException()
Sep 12 18:18:50 nix-bitcoin bash[15886]:   File "/nix/store/6li72ddcdwcdg1m4zdghv2xjw6a72sf7-python3-3.7.6-env/lib/python3.7/site-packages/twisted/python/failure.py", line 488, in raiseException
Sep 12 18:18:50 nix-bitcoin bash[15886]:     raise self.value.with_traceback(self.tb)
Sep 12 18:18:50 nix-bitcoin bash[15886]: twisted.protocols.amp.UnknownRemoteError: Code<UNKNOWN>: Unknown Error
```
when a peer tried to take a yieldgenerator offer. I figure a missing `txtorcon` runtime dependency could have been the issue.

nixbitcoin commented:
The missing `txtorcon` runtime dependency was not the issue. Instead, joinmarket daemon was trying to create `commitmentlist` in the `scripts` directory which was causing a `builtins.OSError: [Errno 30] Read-only file system: 'commitmentlist'`. Fixed in 1e7f845f1df.

I also changed the output of `genwallet.py` and grepped the specific string `recovery_seed` to make sure that seed storage only succeeds when `genwallet.py` really succeeded.

jonasnick commented on modules/joinmarket.nix:
```diff
@@ -0,0 +1,222 @@
+  secretsDir = config.nix-bitcoin.secretsDir;
+
+  torAddress = builtins.head (builtins.split ":" config.services.tor.client.socksListenAddress);
+  configFile = builtins.toFile "config" ''
```
Where's that config from? Perhaps add comment. Do we need to state all these options, or does joinmarket pick the defaults if we don't? In the latter case we should remove these options because then we always have the most up to date defaults.

nixbitcoin replied:
> Where's that config from? Perhaps add comment.

Fixed in `fixup! module: explain where defaultConfig comes from`

> Do we need to state all these options, or does joinmarket pick the defaults if we don't?

AFAICT from https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/configure.py, we do need to state all these options.

jonasnick replied:
Ok

jonasnick commented on modules/joinmarket.nix:
```diff
@@ -0,0 +1,222 @@
+  config = mkIf cfg.enable (mkMerge [{
+    environment.systemPackages = [
+      (hiPrio cfg.cli)
+      pkgs.screen
```
That's up to the user's preferences. I use tmux instead of screen, and I don't know anymore how to use screen, so I wouldn't want it installed.

nixbitcoin replied:
Fixed in `fixup! module: remove screen from systemPackages`

jonasnick replied:
cool

jonasnick reviewed (commented):
I skimmed through the changes and they look good mostly. My issue with this PR is that we should avoid working around things that should be fixed in joinmarket itself and we shouldn't deviate too much from normal usage of joinmarket. In particular,

- I don't really see what the point of the genwallet script is. It means that we're not consistent with joinmarket's usage guide (https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/docs/USAGE.md#using-the-wallet-toolpy-script), requires custom usage instructions and keeping them up-to-date.
- The usage documentation should not do much more than describing the diff to normal joinmarket usage, i.e. the script renamings, how to use the yield generator, etc.
- We can not automatically remove the wallet lock. That's potentially dangerous. If that's a joinmarket problem, it should be fixed upstream

nixbitcoin commented:
> I don't really see what the point of the genwallet script is. It means that we're not consistent with joinmarket's usage guide (https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/docs/USAGE.md#using-the-wallet-toolpy-script), requires custom usage instructions and keeping them up-to-date.

The point of the genwallet script is to automatically create a JoinMarket wallet for ease of use and testing. This allows us to easily use yieldgenerator with a `generate-secrets` generated password, and store & backup the wallet's seed. It also allows us to test an essential function of JoinMarket, wallet generation, easily in the CI.

> The usage documentation should not do much more than describing the diff to normal joinmarket usage, i.e. the script renamings, how to use the yield generator, etc.

Fixed in `fixup! docs: describe diff to normal joinmarket usage`

> We can not automatically remove the wallet lock. That's potentially dangerous. If that's a joinmarket problem, it should be fixed upstream

Fixed in `fixup! module: don't automatically remove wallet lock`. @erikarvstedt Do you ACK?

jonasnick commented on examples/configuration.nix:
```diff
@@ -172,6 +172,15 @@
   # services.backups.with-bulk-data = true;
 
+  ### JOINMARKET
+  # Enable this module to allow using JoinMarket's user interactive scripts (including
```
I'd mention here that joinmarket has priviledged bitcoind access.

nixbitcoin replied:
Implemented in `fixup! module: note joinmarket's full access to bitcoind`

jonasnick reviewed (commented):
Looks good mod nit. Feel free to squash.

nixbitcoin commented:
Squashed all fixups

jonasnick reviewed (approved):
ACK dd882753e6e7782372980c75e9f537904eab6c4e
